### PR TITLE
Initial work

### PR DESCRIPTION
**Pull request by JayH5**
- Working system that reads a set of Marathon JSON group definitions
  and tries to get the state of Marathon to match those definitions.
- This is achieved by simply deleting unknown apps and PUTting the
  group definitions (updating the apps is left up to Marathon).
- There are very basic tests
- TODO: read me
- TODO: Travis, etc

### CODE REVIEW

File: marathon_sync/main.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
+import json
+import treq
+
+from twisted.python import log
+from twisted.web import client
+# Twisted's fault HTTP11 client factory is way too verbose
```
Comment (jerith) [inline marathon_sync/main.py]: Do you mean "default"?

Comment (JayH5) [inline marathon_sync/main.py]: I think so. Copy-pasta-ed from Consular :stuck_out_tongue:

File: marathon_sync/main.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    Group IDs can be specified with or without leading and trailing '/'s. We
+    want a leading '/' but not a trailing one.
+    """
+    if not group_id.startswith('/'):
+        group_id = '/' + group_id
+
+    if group_id.endswith('/') and group_id != '/':
+        group_id = group_id[:-1]
+
+    return group_id
```
Comment (jerith) [inline marathon_sync/main.py]: Would it not be simpler to just strip all leading and trailing slashes and then prepend a leading slash?

Something like `return '/' + group_id.strip('/')` perhaps.

File: setup.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+      author='Jamie Hewland',
+      author_email='jhewland@gmail.com',
+      url='http://github.com/praekelt/marathon-sync',
+      license='BSD',
+      keywords='marathon,mesos',
+      packages=find_packages(exclude=['docs']),
+      include_package_data=True,
+      zip_safe=False,
+      install_requires=requires,
+      tests_require=requires,
```
Comment (jerith) [inline setup.py]: Better to list the requirements here and have `-e .` or whatever in `requirements.txt`. Among other things, requirements files and setup.py support slightly different sets of features and syntax.

**Comment by jerith** [comment]
It's probably worth splitting a Marathon API client out of this at some point, but that's out of scope for this PR.

**Comment by jerith** [comment]
:+1: after making the `setup.py` change. (No need for another review if you test it locally to make sure it works.)

**Comment by smn** [comment]
@jerith, @miltontony & @JayH5 can you guys find like 30m together to talk about this and Mission Control? I think we're largely solving similar problems here and it may be worthwhile collaborating where possible. Even if it's lower level app syncing tools.

**Comment by jerith** [comment]
@smn: This is explicitly a very simple "tell Marathon it should be running exactly the things specified in the config", so the overlap with Mission Control is pretty much just the low-level API call stuff. Happy to discuss it, though. :-)

**Comment by JayH5** [comment]
@smn: sure thing. I'm always keen to prevent duplicated effort.

**Comment by JayH5** [comment]
@jerith there are already a couple of Marathon clients out there. Our use of the API is so simple I thought I'd follow what Consular/Mission Control have done and just make the requests myself.

**Comment by jerith** [comment]
:+1:

**Pull request merged**